src/Sarif.Multitool.Library/Rules/GH1007.FlattenResultMessage.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+using Microsoft.Json.Pointer;
+
+namespace Microsoft.CodeAnalysis.Sarif.Multitool.Rules
+{
+    public class FlattenResultMessage : SarifValidationSkimmerBase
+    {
+        public override ReportingConfiguration DefaultConfiguration => new ReportingConfiguration
+        {
+            Level = FailureLevel.Error,
+            Enabled = this.EnabledByDefault
+        };
+
+        /// <summary>
+        /// GH1007
+        /// </summary>
+        public override string Id => RuleId.FlattenResultMessage;
+
+        public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.GH1007_FlattenResultMessage_FullDescription_Text };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[] {
+            nameof(RuleResources.GH1007_FlattenResultMessage_Error_Default_Text)
+        };
+
+        public override bool EnabledByDefault => false;
+
+        protected override void Analyze(Result result, string resultPointer)
```

Comment:
>Analyze [](http://example.com/codeflow?start=32&length=7)

So now we can't run an analysis without producing errors? Either we recommend using format string args or we object to the absence of flattened messages?

Ideally, the conflicting rules would turn off if the other one is being enforced.

Reply:
how about adding a Set<string> property to `Skimmer` class, e.g. `ConflictSkimmerSet`. In `AnalyzeCommandBase.AnalyzeTarget()` method, iterates available skimmer and add the `ConflictSkimmerSet` value to `disabledSkimmers` set if has value.

this way the rule author can override existing old rule conflicts with the new rule.

what do you think?

Reply:
Interesting! And so the skimmers just need a dependency on rule ids it replaces. Yes, I think this could work.

Reply:
OK, so I thought about your idea a bit more.

Rules should be able to declare an 'IncompatibleRuleIds' set or something similar, All the rule does is to publish this data.

Then we update the analysis engine, which, when instantiating skimmers, makes sure that no new enabled skimmers have a rule id that conflicts with the set of loaded skimmers so far. The analysis engine could raise an error-level configuration notification and exit.

This approach would require users to properly configure analysis to account for the configuration problem. I think everything would drop in pretty cleanly by doing this, we need a low-level update to rules metadata and a simple update to the default skimmer loading behavior.

Reply:
thanks created the issue #2591

regarding "The analysis engine could raise an error-level configuration notification and exit.", should we raise an configuration notification error and disable the incompatible rules then continue the analysis instead exiting?

am thinking about the web validator scenario, user can run only SARIF validator rules, or all SARIF + GH validator rules, if adding a new GH rule which is not compatible with a SARIF rule, we may raise the configuration error and continue analysis without a config file change. If exits right away, the web validator stops getting results until config file is fixed.